google/cloud/storage/idempotency_policy.h

```diff
@@ -0,0 +1,290 @@
+  virtual bool IsIdempotent(
+      internal::PatchObjectRequest const& request) const = 0;
+  virtual bool IsIdempotent(
+      internal::ComposeObjectRequest const& request) const = 0;
+  virtual bool IsIdempotent(
+      internal::RewriteObjectRequest const& request) const = 0;
+  //@}
+
+  //@{
+  /// @name BucketAccessControls resource operations
+  virtual bool IsIdempotent(
+      internal::ListBucketAclRequest const& request) const = 0;
+  virtual bool IsIdempotent(
+      internal::CreateBucketAclRequest const& request) const = 0;
+  virtual bool IsIdempotent(
+      internal::DeleteBucketAclRequest const& request) const = 0;
+  virtual bool IsIdempotent(
+      internal::GetBucketAclRequest const& request) const = 0;
+  virtual bool IsIdempotent(
+      internal::UpdateBucketAclRequest const& request) const = 0;
+  virtual bool IsIdempotent(
+      internal::PatchBucketAclRequest const& request) const = 0;
+  //@}
+
+  //@{
+  /// @name ObjectAccessControls operations
+  virtual bool IsIdempotent(
+      internal::ListObjectAclRequest const& request) const = 0;
+  virtual bool IsIdempotent(
+      internal::CreateObjectAclRequest const& request) const = 0;
+  virtual bool IsIdempotent(
+      internal::DeleteObjectAclRequest const& request) const = 0;
+  virtual bool IsIdempotent(
+      internal::GetObjectAclRequest const& request) const = 0;
+  virtual bool IsIdempotent(
+      internal::UpdateObjectAclRequest const& request) const = 0;
+  virtual bool IsIdempotent(
+      internal::PatchObjectAclRequest const& request) const = 0;
+  //@}
+
+  //@{
+  /// @name DefaultObjectAccessControls operations.
+  virtual bool IsIdempotent(
+      internal::ListDefaultObjectAclRequest const& request) const = 0;
+  virtual bool IsIdempotent(
+      internal::CreateDefaultObjectAclRequest const& request) const = 0;
+  virtual bool IsIdempotent(
+      internal::DeleteDefaultObjectAclRequest const& request) const = 0;
+  virtual bool IsIdempotent(
+      internal::GetDefaultObjectAclRequest const& request) const = 0;
+  virtual bool IsIdempotent(
+      internal::UpdateDefaultObjectAclRequest const& request) const = 0;
+  virtual bool IsIdempotent(
+      internal::PatchDefaultObjectAclRequest const& request) const = 0;
+  //@}
+
+  //@{
+  virtual bool IsIdempotent(
+      internal::GetProjectServiceAccountRequest const& request) const = 0;
+  //@}
+
+  //@{
+  virtual bool IsIdempotent(
+      internal::ListNotificationsRequest const& request) const = 0;
+  virtual bool IsIdempotent(
+      internal::CreateNotificationRequest const& request) const = 0;
+  virtual bool IsIdempotent(
+      internal::GetNotificationRequest const& request) const = 0;
+  virtual bool IsIdempotent(
+      internal::DeleteNotificationRequest const& request) const = 0;
+  //@}
+};
+
+/**
+ * A IdempotencyPolicy that always retries all requests.
+ */
+class AlwaysRetryIdempotencyPolicy : public IdempotencyPolicy {
```

Comment:
I really wish there was a less-bloaty way of doing this, but I don't know of one :/

Reply:
Me too.  The alternative design would be to use a template member function:

```C++
class AlwaysRetryIdempotencyPolicy { public:
  template <typename RequestType> bool IsIdempotent(RequestType const& request) const {
    return true;
  }
};

class StrictIdempotencyPolicy { public:
  template <typename RequestType> bool IsIdempotent(RequestType const& request) const {
    return true;
  }
 bool IsIdempotent(internal::DeleteObjectRequest const& request) const {
    /* do something clever with the preconditions */
 }
};

```

but as you know template member functions cannot be virtual, so we would have to rely on external polymorphism (make the `IdempotencyPolicy` a template parameter of the `RetryPolicy`, so one relies on overload resolution instead of virtual functions).  But I think that would leak the templates everywhere and make this harder to use.

We can change this if it looks terrible though.